android/modules/app/src/java/ti/modules/titanium/app/AppModule.java

```diff
@@ -213,14 +217,23 @@ public boolean getAccessibilityEnabled()
 	@Kroll.method(name = "_restart")
 	public void restart()
 	{
-		Application app = (Application) KrollRuntime.getInstance().getKrollApplication();
-		Intent i = app.getPackageManager().getLaunchIntentForPackage(app.getPackageName());
-		i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
-		i.addFlags(Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED);
-		i.addCategory(Intent.CATEGORY_LAUNCHER);
-		i.setAction(Intent.ACTION_MAIN);
+		KrollRuntime runtime = KrollRuntime.getInstance();
+
+		// prevent termination of root activity via TiBaseActivity.shouldFinishRootActivity()
+		TiBaseActivity.canFinishRoot = false;
+
+		// terminate all activities excluding root
 		TiApplication.terminateActivityStack();
-		app.startActivity(i);
+
+		// allow termination again
+		TiBaseActivity.canFinishRoot = true;
+
+		// restart kroll runtime
+		runtime.doDispose();
+		runtime.initRuntime();
+
+		// manually re-launch app
+		runtime.doRunModule(KrollAssetHelper.readAsset(APP_PATH), APP_PATH, getActivityProxy());
```

Comment:
Our `TiLaunchActivity.loadActivityScript()` and the `resolveUrl()` method it calls does some extra logic when loading a JavaScript on startup. Particularly with Alloy. We might need to do a code shift.
https://github.com/appcelerator/titanium_mobile/blob/master/android/titanium/src/java/org/appcelerator/titanium/TiLaunchActivity.java#L120

Reply:
Ah yes, I'll test these changes with an Alloy app

Reply:
Update, works fine with Alloy apps